src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoder.cs

```diff
@@ -0,0 +1,271 @@
+                                    idx += utf8BytesConsumedForScalar;
+                                }
+                            }
+                        }
+                        else
+                        {
+                            if (DoesAsciiNeedEncoding(ptr[idx]) == 1
+
+                                || DoesAsciiNeedEncoding(ptr[++idx]) == 1
+                                || DoesAsciiNeedEncoding(ptr[++idx]) == 1
+                                || DoesAsciiNeedEncoding(ptr[++idx]) == 1
+                                || DoesAsciiNeedEncoding(ptr[++idx]) == 1
+
+                                || DoesAsciiNeedEncoding(ptr[++idx]) == 1
+                                || DoesAsciiNeedEncoding(ptr[++idx]) == 1
+                                || DoesAsciiNeedEncoding(ptr[++idx]) == 1
+                                || DoesAsciiNeedEncoding(ptr[++idx]) == 1
+
+                                || DoesAsciiNeedEncoding(ptr[++idx]) == 1
+                                || DoesAsciiNeedEncoding(ptr[++idx]) == 1
+                                || DoesAsciiNeedEncoding(ptr[++idx]) == 1
+                                || DoesAsciiNeedEncoding(ptr[++idx]) == 1
+
+                                || DoesAsciiNeedEncoding(ptr[++idx]) == 1
+                                || DoesAsciiNeedEncoding(ptr[++idx]) == 1
+                                || DoesAsciiNeedEncoding(ptr[++idx]) == 1)
+                            {
+                                goto Return;
+                            }
+                            idx++;
+                        }
+                        startingAddress = (sbyte*)ptr + idx;
+                    }
+
+                    // Process the remaining bytes.
+                    Debug.Assert(utf8Text.Length - idx < 16);
+                }
+#endif
+
+                while (idx < utf8Text.Length)
+                {
+                    Debug.Assert((ptr + idx) <= (ptr + utf8Text.Length));
+
+                    if (UnicodeUtility.IsAsciiCodePoint(ptr[idx]))
+                    {
+                        if (DoesAsciiNeedEncoding(ptr[idx]) == 1)
+                        {
+                            goto Return;
+                        }
+                        idx++;
+                    }
+                    else
+                    {
+                        OperationStatus opStatus = UnicodeHelpers.DecodeScalarValueFromUtf8(utf8Text.Slice(idx), out uint nextScalarValue, out int utf8BytesConsumedForScalar);
+
+                        Debug.Assert(nextScalarValue <= int.MaxValue);
+                        if (opStatus != OperationStatus.Done || WillEncode((int)nextScalarValue))
+                        {
+                            goto Return;
+                        }
+
+                        Debug.Assert(opStatus == OperationStatus.Done);
+                        idx += utf8BytesConsumedForScalar;
+                    }
+                }
```

Comment:
does it make sense to add assert after this line idx == utf8Text.Length?

Reply:
The while loop condition already guards for `idx < utf8Text.Length`, so adding the assert would be primarily to make sure idx isn't greater than utf8Text.Length. Sure, we can add it.

Reply:
Addressed this nit in https://github.com/dotnet/corefx/pull/42064